File: tests/r_parity/77_ddd.py
"""StatsPAI triple-differences parity (Python side) -- Module 77.

Runs ``sp.ddd_heterogeneous`` on a staggered DDD panel; the companion R
script runs ``triplediff::ddd`` (Ortiz-Villavicencio & Sant'Anna 2025) on
the same dumped CSV with ``xformla = ~1``, ``est_method = "dr"`` and
never-treated controls. With no covariates the doubly-robust DDD collapses
to the unconditional cell means StatsPAI computes, so the per-``(g, t)``
estimates should agree exactly.

Two things are deliberately NOT compared:

* Standard errors. ``triplediff`` reports analytical influence-function
  SEs; ``sp.ddd_heterogeneous`` has only a cluster bootstrap. Comparing
  them would compare two different variance estimators.
* Pre-treatment placebo cells. ``triplediff`` with ``base_period =
  "varying"`` also reports ``(g, t)`` for ``t < g``; StatsPAI only builds
  post cells.

The overall aggregate is compared under ``weight_by="cohort"``, which is
``triplediff``'s ``pg`` convention (cohort share over all units, both
subgroups). StatsPAI's default weights by treated-eligible units instead;
both are emitted so the harness records the gap rather than hiding it.

Tolerance: rel < 1e-6 on the cell estimates and on the cohort-weighted
aggregate.
"""

from __future__ import annotations

import warnings

import numpy as np
import pandas as pd

from _common import ParityRecord, dump_csv, write_results

MODULE = "77_ddd"
COHORTS = [0, 2, 3, 4]  # 0 = never treated
PERIODS = [1, 2, 3, 4]
COVARS = ["cov1", "cov2"]


def _panel() -> pd.DataFrame:
    """Staggered DDD panel; only the affected subgroup carries the effect.

    The affected share varies across cohorts (0.7 / 0.5 / 0.35) on purpose:
    that is exactly the case where the two aggregation conventions differ,
    so the harness pins the gap rather than a coincidence.
    """
    rng = np.random.default_rng(11)
    share = {2: 0.70, 3: 0.50, 4: 0.35, 0: 0.55}
    rows = []
    uid = 0
    for g in COHORTS:
        for _ in range(150):
            uid += 1
            b = int(rng.random() < share[g])
            fe = rng.normal(0.0, 1.0)
            # Covariates that shift both selection and the outcome path, so
            # the conditional and unconditional estimands genuinely differ.
            c1 = rng.normal(0.4 * b, 1.0)
            c2 = rng.normal(-0.2 * (g != 0), 1.0)
            for t in PERIODS:
                on = g != 0 and t >= g
                effect = (2.0 + 0.5 * (t - g)) if (on and b == 1) else 0.0
                # A subgroup-specific trend the DDD is supposed to net out.
                y = fe + 0.30 * t + 0.45 * b * t + effect + rng.normal(0, 0.8)
                y += (0.5 * c1 - 0.3 * c2) * t
                rows.append(
                    {
                        "id": uid,
                        "time": t,
                        "state": g,
                        "partition": b,
                        "y": y,
                        "cov1": c1,
                        "cov2": c2,
                    }
                )
    return pd.DataFrame(rows)


def main() -> None:
    import statspai as sp

    df = _panel()
    dump_csv(df, MODULE)

    rows: list[ParityRecord] = []
    fits = {}
    for weight_by in ("cohort", "eligible"):
        with warnings.catch_warnings():
            warnings.simplefilter("ignore")
            fits[weight_by] = sp.ddd_heterogeneous(
                df,
                y="y",
                unit="id",
                time="time",
                cohort="state",
                subgroup="partition",
                n_boot=0,
                seed=0,
                weight_by=weight_by,
            )

    for _, row in fits["cohort"].detail.iterrows():
        rows.append(
            ParityRecord(
                module=MODULE,
                side="py",
                statistic=f"ddd_g{int(row['cohort'])}_t{int(row['time'])}",
                estimate=float(row["ddd"]),
                n=int(len(df)),
            )
        )

    rows.append(
        ParityRecord(
            module=MODULE,
            side="py",
            statistic="simple_ATT_cohort_weights",
            estimate=float(fits["cohort"].estimate),
            n=int(len(df)),
        )
    )
    rows.append(
        ParityRecord(
            module=MODULE,
            side="py",
            statistic="simple_ATT_eligible_weights",
            estimate=float(fits["eligible"].estimate),
            n=int(len(df)),
        )
    )

    # Conditional DDD: all three nuisance combinations, cells and analytic
    # standard errors. The SEs are compared here (unlike the unconditional
    # block above) because the analytic path IS the reference's variance
    # estimator, not a bootstrap standing in for it.
    for method in ("dr", "ipw", "reg"):
        with warnings.catch_warnings():
            warnings.simplefilter("ignore")
            cond = sp.ddd_heterogeneous(
                df,
                y="y",
                unit="id",
                time="time",
                cohort="state",
                subgroup="partition",
                n_boot=0,
                seed=0,
                weight_by="cohort",
                x=COVARS,
                est_method=method,
                se="analytic",
            )
        for _, row in cond.detail.iterrows():
            tag = f"{method}_g{int(row['cohort'])}_t{int(row['time'])}"
            rows.append(
                ParityRecord(
                    module=MODULE,
                    side="py",
                    statistic=f"ddd_{tag}",
                    estimate=float(row["ddd"]),
                    se=float(row["se"]),
                    n=int(len(df)),
                )
            )
        rows.append(
            ParityRecord(
                module=MODULE,
                side="py",
                statistic=f"simple_ATT_{method}",
                estimate=float(cond.estimate),
                se=float(cond.se),
                n=int(len(df)),
            )
        )

    write_results(MODULE, "py", rows, extra={"estimator": "ddd_heterogeneous"})


if __name__ == "__main__":
    main()

File: tests/r_parity/77_ddd.R
# StatsPAI triple-differences parity (R side) -- Module 77.
#
# Runs triplediff::ddd (Ortiz-Villavicencio & Sant'Anna 2025) on the panel
# dumped by 77_ddd.py, with no covariates so the doubly-robust DDD reduces
# to the unconditional cell means sp.ddd_heterogeneous computes, and
# agg_ddd(type = "simple") for the overall number.
#
# Only post-treatment cells are emitted: base_period = "varying" also
# produces (g, t) for t < g, which the Python side does not build.
#
# Standard errors are not compared -- triplediff reports analytical
# influence-function SEs and sp.ddd_heterogeneous only has a cluster
# bootstrap.
#
# Tolerance: rel < 1e-6.

.args <- commandArgs(trailingOnly = FALSE)
.file_arg <- grep("^--file=", .args, value = TRUE)
.script_dir <- if (length(.file_arg) > 0) {
  dirname(normalizePath(sub("^--file=", "", .file_arg[1])))
} else {
  getwd()
}
source(file.path(.script_dir, "_common.R"))

suppressPackageStartupMessages({
  library(triplediff)
  library(data.table)
})

MODULE <- "77_ddd"

df <- as.data.table(read_csv_strict(MODULE))

res <- ddd(
  yname = "y", tname = "time", idname = "id", gname = "state",
  pname = "partition", xformla = ~1, data = df,
  control_group = "nevertreated", base_period = "varying",
  est_method = "dr", panel = TRUE, boot = FALSE
)

rows <- list()
post <- res$periods >= res$groups
for (k in which(post)) {
  rows[[length(rows) + 1]] <- parity_row(
    module    = MODULE,
    statistic = sprintf("ddd_g%d_t%d", res$groups[k], res$periods[k]),
    estimate  = res$ATT[k],
    se        = NA_real_,
    n         = nrow(df)
  )
}

agg <- agg_ddd(res, type = "simple")$aggte_ddd
rows[[length(rows) + 1]] <- parity_row(
  module    = MODULE,
  statistic = "simple_ATT_cohort_weights",
  estimate  = agg$overall.att,
  se        = NA_real_,
  n         = nrow(df)
)

# Conditional DDD across the three nuisance combinations. Standard errors
# ARE compared here: sp.ddd_heterogeneous(se="analytic") uses the same
# influence-function variance, so a gap would be a real disagreement rather
# than two different estimators of the same quantity.
for (method in c("dr", "ipw", "reg")) {
  cond <- ddd(
    yname = "y", tname = "time", idname = "id", gname = "state",
    pname = "partition", xformla = ~ cov1 + cov2, data = df,
    control_group = "nevertreated", base_period = "varying",
    est_method = method, panel = TRUE, boot = FALSE
  )
  post_c <- cond$periods >= cond$groups
  for (k in which(post_c)) {
    rows[[length(rows) + 1]] <- parity_row(
      module    = MODULE,
      statistic = sprintf("ddd_%s_g%d_t%d", method, cond$groups[k], cond$periods[k]),
      estimate  = cond$ATT[k],
      se        = cond$se[k],
      n         = nrow(df)
    )
  }
  agg_c <- agg_ddd(cond, type = "simple")$aggte_ddd
  rows[[length(rows) + 1]] <- parity_row(
    module    = MODULE,
    statistic = sprintf("simple_ATT_%s", method),
    estimate  = agg_c$overall.att,
    se        = agg_c$overall.se,
    n         = nrow(df)
  )
}

write_results(MODULE, rows, extra = list(
  reference = "triplediff::ddd + agg_ddd(type='simple')",
  triplediff = as.character(utils::packageVersion("triplediff"))
))
